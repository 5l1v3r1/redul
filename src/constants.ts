export const TEXT_ELEMENT_TYPE = '@@reax/text-node'
export const ROOT_FIBER_NODE = '@@reax/root-fiber-node'

export const ENOUGH_TIME = 1
