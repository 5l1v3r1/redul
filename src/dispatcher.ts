import { FiberNode } from '../reax'

const dispatcher = {
    render: (rootFiberNode: Reax.FiberNode) => {
        // console.log(rootFiberNode)
    }
}

export default dispatcher
