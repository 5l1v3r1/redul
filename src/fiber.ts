import { FiberNode, ElementInput, RootHTMLElementWithFiberNode } from "../reax";
import { FiberNodeTag } from '../interface'
import { transformElementInputsToElements } from "./element";
import { ROOT_FIBER_NODE } from "./constants";

export function createRootFiberNode(element: ElementInput | ElementInput[], statNode: RootHTMLElementWithFiberNode): FiberNode {
    return {
        tag: FiberNodeTag.HOST_ROOT_NODE,
        children: transformElementInputsToElements(element),
        effects: [],
        statNode,
        alternate: statNode[ROOT_FIBER_NODE] || null
    }
}
