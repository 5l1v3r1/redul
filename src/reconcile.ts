import { createRootFiberNode } from './fiber'
import { ENOUGH_TIME, ROOT_FIBER_NODE } from './constants'
import dispatcher from './dispatcher'
import { isComponent } from './utils'
import { transformElementInputsToElements } from './element'
import { FiberNode, ElementInput, FunctionComponent, Element, RootHTMLElementWithFiberNode } from '../reax'
import { EffectTag, FiberNodeTag } from '../interface'

const taskQueue: FiberNode[] = []

const requestIdleCallback = (callback: (deadline: RequestIdleCallbackDeadline) => void) => {
    callback({
        didTimeout: true,
        timeRemaining: () => 100
    })
}
let nextUnitWork: FiberNode | null = null
let currentRootFiberNode: FiberNode | null = null

export function render(element: ElementInput, containerDom: HTMLElement) {
    // clear all before render
    dispatcher.clearDomContent(containerDom)
    const rootFiberNode = createRootFiberNode(element, containerDom)
    currentRootFiberNode = rootFiberNode
    taskQueue.push(rootFiberNode)

    requestIdleCallback(performWork)
    return containerDom
}

function performWork(deadline: RequestIdleCallbackDeadline) {
    nextUnitWork = resolveNextUnitWork()

    if (!nextUnitWork) {
        commitAllWork()
        return
    }

    if (deadline.timeRemaining() > ENOUGH_TIME) {
        nextUnitWork = performUnitWork(nextUnitWork)
    }

    requestIdleCallback(performWork)
}

function resolveNextUnitWork() {
    nextUnitWork = nextUnitWork || taskQueue.shift() || null
    return nextUnitWork
}

function commitAllWork() {
    if (currentRootFiberNode) {
        // save root fiber
        (currentRootFiberNode.statNode as RootHTMLElementWithFiberNode)[ROOT_FIBER_NODE] = currentRootFiberNode
        dispatcher.render(currentRootFiberNode)
    }
}

function performUnitWork(unitWork: FiberNode) {
    let fiberNode: FiberNode | null = unitWork
    beginUnitWork(fiberNode)

    if (fiberNode.child) {
        return fiberNode.child
    }

    while(fiberNode) {
        completeUnitWork(fiberNode)
        if (fiberNode.sibling) {
            return fiberNode.sibling
        }
        fiberNode = fiberNode.parent || null
    }

    return null
}

function beginUnitWork(fiberNode: FiberNode) {
    if (isComponent(fiberNode.type)) {
        beginComponentNodeUnitWork(fiberNode)
    } else (
        beginHostNodeUnitWork(fiberNode)
    )
}

function beginComponentNodeUnitWork(fiberNode: FiberNode) {
    const Component = fiberNode.type as FunctionComponent
    const children = transformElementInputsToElements(Component(fiberNode.props))
    reconcileChildren(children, fiberNode)
}

function beginHostNodeUnitWork(fiberNode: FiberNode) {
    const { children } = fiberNode
    reconcileChildren(children, fiberNode)
}

function reconcileChildren(children: Element[], fiberNode: FiberNode) {
    const alternateParentFiberNode = fiberNode.alternate
    // fiber node chain
    let prevChildFiberNode: FiberNode | null = null
    let alternateChildFiberNode: FiberNode | null = null
    for (let i = 0; i < children.length; i++) {
        const childElement = children[i]
        const childFiberNode = transformElementToFiberNode(childElement)
        if (i === 0) {
            fiberNode.child = childFiberNode
            alternateChildFiberNode = alternateParentFiberNode && alternateParentFiberNode.child || null
        } else {
            prevChildFiberNode!.sibling = childFiberNode
            alternateChildFiberNode = alternateChildFiberNode && alternateChildFiberNode.sibling || null
        }
        childFiberNode.parent = fiberNode
        childFiberNode.alternate = alternateChildFiberNode
        childFiberNode.statNode = alternateChildFiberNode && alternateChildFiberNode.statNode || null
        childFiberNode.effectTag = resolveFiberNodeEffectTag(childElement, alternateChildFiberNode)

        prevChildFiberNode = childFiberNode
    }

    if (alternateChildFiberNode && alternateChildFiberNode.sibling) {
        resolveAlternateFiberNodesAsRemoveEffectTag(alternateChildFiberNode.sibling, fiberNode)
    }
}

function transformElementToFiberNode(element: Element): FiberNode {
    const fiberNode = {
        tag: isComponent(element.type) ? FiberNodeTag.COMPONENT_NODE : FiberNodeTag.HOST_NODE,
        ...element,
        effects: [],
        statNode: null
    }
    return fiberNode
}

function resolveFiberNodeEffectTag(element: Element, alternateFiberNode: FiberNode | null): EffectTag {
    if (alternateFiberNode) {
        if (element.type === alternateFiberNode.type) {
            return EffectTag.UPDATE
        } else {
            return EffectTag.REPLACE
        }
    }

    return EffectTag.ADD
}

function resolveAlternateFiberNodesAsRemoveEffectTag(alternateFiberNode: FiberNode | null, fiberNode: FiberNode) {
    if (alternateFiberNode) {
        alternateFiberNode.effectTag = EffectTag.REMOVE
        fiberNode.effects.push(alternateFiberNode)
    }
}

function completeUnitWork(fiberNode: FiberNode) {
    const parentFiberNode = fiberNode.parent
    const effects = fiberNode.effects
    if (parentFiberNode) {
        if (fiberNode.effectTag) {
            parentFiberNode.effects.push(fiberNode)
        }
        parentFiberNode.effects.push(...effects)
    }
}
