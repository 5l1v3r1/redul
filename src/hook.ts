import { Hook, FiberNode } from "../reax";
import { update } from './reconcile'

// hook {
//     memoziedState, ✅
//     baseState, ✅
//     queue,
//     baseUpdate,
//     next, ✅
// }

let workInProgressHook: Hook | null = null
let workInProgressFiberNode: FiberNode | null = null

export function setWorkInProgressFiberNode(fiberNode: FiberNode) {
    workInProgressFiberNode = fiberNode
}

export function resetWorkInProgressHook() {
    workInProgressHook = null
}

function mountWorkInProgressHook<S>() {
    const hook: Hook<S> = {
        memoizedState: null,
        dispatch: null,
        next: null
    }

    if (workInProgressHook === null) {
        // first hook of hook link list
        workInProgressHook = hook
        // save in work-in-progress fiber
        if (workInProgressFiberNode) {
            workInProgressFiberNode.hooks = hook
        }
    } else {
        workInProgressHook.next = hook
    }

    return hook
}

function updateWorkInProgressHook<S>() {
    // hooks never be null
    let hook: Hook<S> | null = null
    if (workInProgressHook === null) {
        hook = workInProgressFiberNode!.hooks!
    } else {
        hook = workInProgressHook.next!
    }

    return hook
}

function isInitStateFunc<S>(initState: S | (() => S)): initState is () => S {
    return typeof initState === 'function'
}

function mountUseState<S>(initState: S | (() => S)): [S, (newState: S) => void] {
    const hook = mountWorkInProgressHook<S>()
    hook.memoizedState = isInitStateFunc(initState) ? initState() : initState
    hook.next = hook
    hook.dispatch = (newState: S) => dispatchAction(hook, workInProgressFiberNode, newState)
    return [hook.memoizedState, hook.dispatch]
}

function updateUseState<S>(initState: S | (() => S)): [S | null, ((newState: S) => void) | null] {
    const hook = updateWorkInProgressHook<S>()
    return [hook.memoizedState, hook.dispatch]
}

function dispatchAction<S>(hook: Hook<S>, fiberNode: FiberNode | null, newState: S) {
    if (fiberNode) {
        hook.memoizedState = newState
        update(fiberNode)
    }
}

function useState<S>(initState: S) {
    if (workInProgressFiberNode && workInProgressFiberNode.hooks) {
        return updateUseState(initState)
    }

    return mountUseState(initState)
}


export {
    useState
}
