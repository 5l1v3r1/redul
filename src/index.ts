import { createElement } from './element'
import { render } from './reconcile'
import { useState, useReducer, useEffect, useCallback, useMemo, useRef } from './hook'

export {
    createElement,
    render,
    useState,
    useReducer,
    useEffect,
    useCallback,
    useMemo,
    useRef
}
