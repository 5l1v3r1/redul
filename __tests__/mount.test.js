import { render, createElement as h } from '../src'

const getRootNode = () => {
    document.body.innerHTML = '<div id="root"></div>'
    const $root = document.getElementById('root')
    return $root
}
const $root = getRootNode()

describe('fiber reconcile test', () => {
    describe('mount stage', () => {
        test('element string type', () => {

            expect(render(<div className="test"></div>, $root)).toMatchSnapshot()
        })

        test('element function type', () => {
            function Count({ count }) {
                return <div>{count}</div>
            }
            expect(render(<Count count={1} />, $root)).toMatchSnapshot()
        })
    })
})
