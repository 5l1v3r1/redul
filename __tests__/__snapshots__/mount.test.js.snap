// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fiber reconcile test mount stage element function type 1`] = `
Object {
  "alternate": null,
  "child": Object {
    "alternate": null,
    "child": Object {
      "alternate": null,
      "child": Object {
        "alternate": null,
        "children": Array [],
        "effectTag": 1,
        "effects": Array [],
        "parent": [Circular],
        "props": Object {
          "children": Array [],
          "nodeValue": 1,
        },
        "statNode": null,
        "tag": 1,
        "type": "@@reax/text-node",
      },
      "children": Array [
        Object {
          "children": Array [],
          "props": Object {
            "children": Array [],
            "nodeValue": 1,
          },
          "type": "@@reax/text-node",
        },
      ],
      "effectTag": 1,
      "effects": Array [
        Object {
          "alternate": null,
          "children": Array [],
          "effectTag": 1,
          "effects": Array [],
          "parent": [Circular],
          "props": Object {
            "children": Array [],
            "nodeValue": 1,
          },
          "statNode": null,
          "tag": 1,
          "type": "@@reax/text-node",
        },
      ],
      "parent": [Circular],
      "props": Object {
        "children": Array [
          Object {
            "children": Array [],
            "props": Object {
              "children": Array [],
              "nodeValue": 1,
            },
            "type": "@@reax/text-node",
          },
        ],
      },
      "statNode": null,
      "tag": 1,
      "type": "div",
    },
    "children": Array [],
    "effectTag": 1,
    "effects": Array [
      Object {
        "alternate": null,
        "child": Object {
          "alternate": null,
          "children": Array [],
          "effectTag": 1,
          "effects": Array [],
          "parent": [Circular],
          "props": Object {
            "children": Array [],
            "nodeValue": 1,
          },
          "statNode": null,
          "tag": 1,
          "type": "@@reax/text-node",
        },
        "children": Array [
          Object {
            "children": Array [],
            "props": Object {
              "children": Array [],
              "nodeValue": 1,
            },
            "type": "@@reax/text-node",
          },
        ],
        "effectTag": 1,
        "effects": Array [
          Object {
            "alternate": null,
            "children": Array [],
            "effectTag": 1,
            "effects": Array [],
            "parent": [Circular],
            "props": Object {
              "children": Array [],
              "nodeValue": 1,
            },
            "statNode": null,
            "tag": 1,
            "type": "@@reax/text-node",
          },
        ],
        "parent": [Circular],
        "props": Object {
          "children": Array [
            Object {
              "children": Array [],
              "props": Object {
                "children": Array [],
                "nodeValue": 1,
              },
              "type": "@@reax/text-node",
            },
          ],
        },
        "statNode": null,
        "tag": 1,
        "type": "div",
      },
      Object {
        "alternate": null,
        "children": Array [],
        "effectTag": 1,
        "effects": Array [],
        "parent": Object {
          "alternate": null,
          "child": [Circular],
          "children": Array [
            Object {
              "children": Array [],
              "props": Object {
                "children": Array [],
                "nodeValue": 1,
              },
              "type": "@@reax/text-node",
            },
          ],
          "effectTag": 1,
          "effects": Array [
            [Circular],
          ],
          "parent": [Circular],
          "props": Object {
            "children": Array [
              Object {
                "children": Array [],
                "props": Object {
                  "children": Array [],
                  "nodeValue": 1,
                },
                "type": "@@reax/text-node",
              },
            ],
          },
          "statNode": null,
          "tag": 1,
          "type": "div",
        },
        "props": Object {
          "children": Array [],
          "nodeValue": 1,
        },
        "statNode": null,
        "tag": 1,
        "type": "@@reax/text-node",
      },
    ],
    "parent": [Circular],
    "props": Object {
      "children": Array [],
      "count": 1,
    },
    "statNode": null,
    "tag": 3,
    "type": [Function],
  },
  "children": Array [
    Object {
      "children": Array [],
      "props": Object {
        "children": Array [],
        "count": 1,
      },
      "type": [Function],
    },
  ],
  "effects": Array [
    Object {
      "alternate": null,
      "child": Object {
        "alternate": null,
        "child": Object {
          "alternate": null,
          "children": Array [],
          "effectTag": 1,
          "effects": Array [],
          "parent": [Circular],
          "props": Object {
            "children": Array [],
            "nodeValue": 1,
          },
          "statNode": null,
          "tag": 1,
          "type": "@@reax/text-node",
        },
        "children": Array [
          Object {
            "children": Array [],
            "props": Object {
              "children": Array [],
              "nodeValue": 1,
            },
            "type": "@@reax/text-node",
          },
        ],
        "effectTag": 1,
        "effects": Array [
          Object {
            "alternate": null,
            "children": Array [],
            "effectTag": 1,
            "effects": Array [],
            "parent": [Circular],
            "props": Object {
              "children": Array [],
              "nodeValue": 1,
            },
            "statNode": null,
            "tag": 1,
            "type": "@@reax/text-node",
          },
        ],
        "parent": [Circular],
        "props": Object {
          "children": Array [
            Object {
              "children": Array [],
              "props": Object {
                "children": Array [],
                "nodeValue": 1,
              },
              "type": "@@reax/text-node",
            },
          ],
        },
        "statNode": null,
        "tag": 1,
        "type": "div",
      },
      "children": Array [],
      "effectTag": 1,
      "effects": Array [
        Object {
          "alternate": null,
          "child": Object {
            "alternate": null,
            "children": Array [],
            "effectTag": 1,
            "effects": Array [],
            "parent": [Circular],
            "props": Object {
              "children": Array [],
              "nodeValue": 1,
            },
            "statNode": null,
            "tag": 1,
            "type": "@@reax/text-node",
          },
          "children": Array [
            Object {
              "children": Array [],
              "props": Object {
                "children": Array [],
                "nodeValue": 1,
              },
              "type": "@@reax/text-node",
            },
          ],
          "effectTag": 1,
          "effects": Array [
            Object {
              "alternate": null,
              "children": Array [],
              "effectTag": 1,
              "effects": Array [],
              "parent": [Circular],
              "props": Object {
                "children": Array [],
                "nodeValue": 1,
              },
              "statNode": null,
              "tag": 1,
              "type": "@@reax/text-node",
            },
          ],
          "parent": [Circular],
          "props": Object {
            "children": Array [
              Object {
                "children": Array [],
                "props": Object {
                  "children": Array [],
                  "nodeValue": 1,
                },
                "type": "@@reax/text-node",
              },
            ],
          },
          "statNode": null,
          "tag": 1,
          "type": "div",
        },
        Object {
          "alternate": null,
          "children": Array [],
          "effectTag": 1,
          "effects": Array [],
          "parent": Object {
            "alternate": null,
            "child": [Circular],
            "children": Array [
              Object {
                "children": Array [],
                "props": Object {
                  "children": Array [],
                  "nodeValue": 1,
                },
                "type": "@@reax/text-node",
              },
            ],
            "effectTag": 1,
            "effects": Array [
              [Circular],
            ],
            "parent": [Circular],
            "props": Object {
              "children": Array [
                Object {
                  "children": Array [],
                  "props": Object {
                    "children": Array [],
                    "nodeValue": 1,
                  },
                  "type": "@@reax/text-node",
                },
              ],
            },
            "statNode": null,
            "tag": 1,
            "type": "div",
          },
          "props": Object {
            "children": Array [],
            "nodeValue": 1,
          },
          "statNode": null,
          "tag": 1,
          "type": "@@reax/text-node",
        },
      ],
      "parent": [Circular],
      "props": Object {
        "children": Array [],
        "count": 1,
      },
      "statNode": null,
      "tag": 3,
      "type": [Function],
    },
    Object {
      "alternate": null,
      "child": Object {
        "alternate": null,
        "children": Array [],
        "effectTag": 1,
        "effects": Array [],
        "parent": [Circular],
        "props": Object {
          "children": Array [],
          "nodeValue": 1,
        },
        "statNode": null,
        "tag": 1,
        "type": "@@reax/text-node",
      },
      "children": Array [
        Object {
          "children": Array [],
          "props": Object {
            "children": Array [],
            "nodeValue": 1,
          },
          "type": "@@reax/text-node",
        },
      ],
      "effectTag": 1,
      "effects": Array [
        Object {
          "alternate": null,
          "children": Array [],
          "effectTag": 1,
          "effects": Array [],
          "parent": [Circular],
          "props": Object {
            "children": Array [],
            "nodeValue": 1,
          },
          "statNode": null,
          "tag": 1,
          "type": "@@reax/text-node",
        },
      ],
      "parent": Object {
        "alternate": null,
        "child": [Circular],
        "children": Array [],
        "effectTag": 1,
        "effects": Array [
          [Circular],
          Object {
            "alternate": null,
            "children": Array [],
            "effectTag": 1,
            "effects": Array [],
            "parent": [Circular],
            "props": Object {
              "children": Array [],
              "nodeValue": 1,
            },
            "statNode": null,
            "tag": 1,
            "type": "@@reax/text-node",
          },
        ],
        "parent": [Circular],
        "props": Object {
          "children": Array [],
          "count": 1,
        },
        "statNode": null,
        "tag": 3,
        "type": [Function],
      },
      "props": Object {
        "children": Array [
          Object {
            "children": Array [],
            "props": Object {
              "children": Array [],
              "nodeValue": 1,
            },
            "type": "@@reax/text-node",
          },
        ],
      },
      "statNode": null,
      "tag": 1,
      "type": "div",
    },
    Object {
      "alternate": null,
      "children": Array [],
      "effectTag": 1,
      "effects": Array [],
      "parent": Object {
        "alternate": null,
        "child": [Circular],
        "children": Array [
          Object {
            "children": Array [],
            "props": Object {
              "children": Array [],
              "nodeValue": 1,
            },
            "type": "@@reax/text-node",
          },
        ],
        "effectTag": 1,
        "effects": Array [
          [Circular],
        ],
        "parent": Object {
          "alternate": null,
          "child": [Circular],
          "children": Array [],
          "effectTag": 1,
          "effects": Array [
            [Circular],
            [Circular],
          ],
          "parent": [Circular],
          "props": Object {
            "children": Array [],
            "count": 1,
          },
          "statNode": null,
          "tag": 3,
          "type": [Function],
        },
        "props": Object {
          "children": Array [
            Object {
              "children": Array [],
              "props": Object {
                "children": Array [],
                "nodeValue": 1,
              },
              "type": "@@reax/text-node",
            },
          ],
        },
        "statNode": null,
        "tag": 1,
        "type": "div",
      },
      "props": Object {
        "children": Array [],
        "nodeValue": 1,
      },
      "statNode": null,
      "tag": 1,
      "type": "@@reax/text-node",
    },
  ],
  "statNode": <div
    id="root"
  />,
  "tag": 2,
}
`;

exports[`fiber reconcile test mount stage element string type 1`] = `
Object {
  "alternate": null,
  "child": Object {
    "alternate": null,
    "children": Array [],
    "effectTag": 1,
    "effects": Array [],
    "parent": [Circular],
    "props": Object {
      "children": Array [],
      "className": "test",
    },
    "statNode": null,
    "tag": 1,
    "type": "div",
  },
  "children": Array [
    Object {
      "children": Array [],
      "props": Object {
        "children": Array [],
        "className": "test",
      },
      "type": "div",
    },
  ],
  "effects": Array [
    Object {
      "alternate": null,
      "children": Array [],
      "effectTag": 1,
      "effects": Array [],
      "parent": [Circular],
      "props": Object {
        "children": Array [],
        "className": "test",
      },
      "statNode": null,
      "tag": 1,
      "type": "div",
    },
  ],
  "statNode": <div
    id="root"
  />,
  "tag": 2,
}
`;
